py/dml/codegen.py

```diff
@@ -238,6 +245,135 @@ def codegen_exit(self, site, retvals):
+        targets = []
+        for ((name, typ), val) in zip(self.outp, retvals):
+            target = self.mkRef(f'p_{name}', typ)
+            stmts.append(mkCopyData(site, val, target))
+            targets.append(target)
+        stmts.append(codegen_return(site, self.outp, self.throws, targets))
+        return mkCompound(site, stmts)
+
+class Memoization:
+    def prelude(self):
+        pass
+    def exit_handler(self):
+        pass
+    def fail_handler(self):
+        pass
+
+class IndependentMemoized(Memoization):
+    def __init__(self, func):
+        assert func.method.independent
+        self.func = func
+        self.method = func.method
+
+    def mkRef(self, ref, typ):
+        indexing = ''.join([f'[_idx{i}]'
+                            for i in range(self.method.dimensions)])
+        return mkLit(self.method.site, f'_memo{indexing}.{ref}', typ)
+
+    def prelude(self):
+        struct_body = ''.join(
+            f'{typ.declaration(name)}; '
+            for (name, typ) in (
+                    [('ran', TInt(8, True))]
+                    + ([('threw', TBool())] if self.func.throws else [])
+                    + [(f'p_{name}', typ) for (name, typ) in self.func.outp]))
+        array_defs = ''.join([f'[{i}]' for i in self.method.dimsizes])
+        struct_var_def = mkInline(
+            self.method.site,
+            f'static struct {{{struct_body}}} _memo{array_defs};')
+        return [struct_var_def] + memoization_common_prelude(
+            self.method.logname_anonymized(), self.method.site, self.func.outp,
+            self.func.throws, self.mkRef)
+    def exit_handler(self):
+        return MemoizedReturnExit(self.method.site, self.func.outp,
+                                  self.func.throws, self.mkRef)
+    def fail_handler(self):
+        return (MemoizedReturnFailure(self.method.rbrace_site, self.mkRef)
+                if self.func.throws else NoFailure(self.method.site))
+
+class SharedIndependentMemoized(Memoization):
+    def __init__(self, method):
+        assert method.independent
+        self.method = method
+    def mkRef(self, ref, typ):
+        traitname = cident(self.method.trait.name)
+        return mkLit(self.method.site,
+                     f'((struct _{traitname} *) _{traitname}.trait)'
+                     + f'->_{self.method.name}_outs.{ref}', typ)
+    def prelude(self):
+        return memoization_common_prelude(
+            self.method.name, self.method.site, self.method.outp,
+            self.method.throws, self.mkRef)
+    def exit_handler(self):
+        return MemoizedReturnExit(self.method.site, self.method.outp,
+                                  self.method.throws, self.mkRef)
+    def fail_handler(self):
+        return (MemoizedReturnFailure(self.method.rbrace_site, self.mkRef)
+                if self.method.throws else NoFailure(self.method.site))
+
+def memoization_common_prelude(name, site, outp, throws, mkRef):
```

Comment:
Would be nice with a comment that explains `ran` and `threw`, and how they solve memo problems.